fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java

```diff
@@ -27,23 +27,87 @@
  * {@link KeyBindingRegistry#register(FabricKeyBinding)}!
  */
 public class FabricKeyBinding extends KeyBinding {
+
+    private final Identifier id;
+
 	protected FabricKeyBinding(Identifier id, InputUtil.Type type, int code, String category) {
 		super("key." + id.toString().replace(':', '.'), type, code, category);
+		this.id = id;
+	}
+
+	/**
+	 * Original identifier used to register this key.
+	 * May be different from the {@link getId()}!
+	 */
+	public Identifier getIdentifier() {
+	    return id;
 	}
 
 	public static class Builder {
-		protected final FabricKeyBinding binding;
 
-		protected Builder(FabricKeyBinding binding) {
-			this.binding = binding;
+		private InputUtil.Type type = InputUtil.Type.KEYSYM;
+
+		private Identifier id = new Identifier("fabric", "unnamed");
+
+		private int code;
+
+		private String category = KeyCategory.MISC;
+
+		private Builder() {
+
+		}
+
+		public Builder id(Identifier keyName) {
+		    this.id = keyName;
+			return this;
+		}
+
+		public Builder category(String category) {
+			this.category = category;
+			return this;
+		}
+
+		public Builder code(int keyCode) {
+			this.code = keyCode;
+			return this;
+		}
+
+		public Builder type(InputUtil.Type type) {
+			this.type = type;
+			return this;
 		}
 
 		public FabricKeyBinding build() {
-			return binding;
+			return new FabricKeyBinding(id, type, code, category);
```

Comment:
You should be checking if everything has been set first. Throw `IllegalStateExeption` if it hasn't.

Reply:
That might be a good idea. I'll put in checks for the required stuff.

Reply:
If you're going to use `NullPointerException`,  you'd might as well use `Objects.requireNonNull()`

Reply:
Never really used that before...

Reply:
Oh. This is actually part of Java? I always through Objects/Lists/Maps/etc were strictly part of Guava.

Reply:
Guava's version is `Preconditions.checkNotNull()`. Apache uses `Validate.nonNull()`.

Reply:
Weird. i would've avoided them simply for the sake of not depending on those libs.

Reply:
There's some really useful stuff in the default MC libraries, I wouldn't avoid using them, you're just making your life so much harder. The only benefit is that the code would be portable to another program, which this already isn't. Guava in particular is absolutely packed with useful stuff.